### PR TITLE
[SPARK-47366][SQL] Implement parse_json.

### PR DESCRIPTION
**Pull request by chenhao-db**
### What changes were proposed in this pull request?

This PR adds the implementation of the `parse_json` expression to replace the current fake implementation. This expression parses a JSON string as a variant value. It throws an exception when the string is not a valid JSON value or when the resulting variant cannot fit into the size limit.

This feature is achieved by introducing a library`common/variant`. It includes the utility functions to build and manipulate binary-encoded variant values. It also contains a `README` that describes the variant binary format. It is intended that this library can be used outside of Spark.

Some usage examples of `parse_json`:

```
select parse_json('{"a": 1, "b": 2}');

create table variant_table as select parse_json(j) as v from json_table;

select parse_json('[');
-- will throw an exception because the input is not valid JSON

select parse_json('"' || repeat('a', 16 * 1024 * 1024) || '"');
-- will throw an exception because the variant exceeds the size limit
```

### Does this PR introduce _any_ user-facing change?

Yes, the `parse_json` expression will return an actual binary-encoded variant value rather than the original placeholder value.

### How was this patch tested?

Unit tests that validate the `parse_json` result. Negative cases where the expression fail on invalid/large JSON are also covered.

Some unit tests need to be temporarily disabled because the `toString` implementation doesn't match the `parse_json` implementation yet. I will shortly add a new `toString` implementation and re-enable them.

### CODE REVIEW

**Comment by dtenedor** [comment]
@HyukjinKwon @cloud-fan do you want to help look? This approach seems fine to me.

File: project/SparkBuild.scala [hunk 1]
```diff
@@ -58,10 +58,10 @@ object BuildCommons {
 
   val allProjects@Seq(
     core, graphx, mllib, mllibLocal, repl, networkCommon, networkShuffle, launcher, unsafe, tags, sketch, kvstore,
-    commonUtils, sqlApi, _*
+    commonUtils, sqlApi, variant, _*
```
Comment (HyukjinKwon) [inline project/SparkBuild.scala:61]: Would also have to add at `dev/sparktestsupport/modules.py` and `.github/workflows/build_and_test.yml` so it runs the tests for individual modules.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/VariantVal.java [hunk 2]
```diff
@@ -107,4 +107,17 @@ public String toString() {
     // NOTE: the encoding is not yet implemented, this is not the final implementation.
     return new String(value);
   }
+
+  /**
+   * Compare two variants in bytes. The variant equality is more complex than it, and we haven't
+   * supported it in the user surface yet. This method is only intended for tests.
+   */
+  @Override
+  public boolean equals(Object other) {
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/VariantVal.java:116]: `equals` should be overriden with `hashCode` together.

File: common/variant/README.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+Another motivation for the representation is that (aside from metadata) each inner Variant value is contiguous and self-contained. For example, in a Variant containing an Array of Variant values, the representation of an inner Variant value, when paired with the metadata of the full variant, is itself a valid Variant.
+
+# Metadata encoding
+
+The grammar for encoded metadata is as follows
+
+```
+metadata: <header> <dictionary_size> <dictionary>
+header: 1 byte (<version> | <sorted_strings> << 4 | (<offset_size_minus_one> << 6))
+version: a 4-bit version ID. Currently, must always contain the value 1
```
Comment (cloud-fan) [inline common/variant/README.md:23]: hmmm, at most 16 versions?

File: common/variant/src/main/java/org/apache/spark/variant/Variant.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.variant;
+
+/**
+ * This class is structurally equivalent to {@link org.apache.spark.unsafe.types.VariantVal}. We
+ * define a new class to avoid depending on or modifying Spark.
```
Comment (cloud-fan) [inline common/variant/src/main/java/org/apache/spark/variant/Variant.java:22]: shall we make Spark depend on this class and remove `VariantVal`?

**Review by chenhao-db — COMMENTED** [review]
@cloud-fan Thanks for reviewing!

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @chenhao-db, @cloud-fan , @HyukjinKwon .  The following package name sounds a little misleading because it could be interpreted a variant of Apache Spark. Do you think we can revise this package more specifically?

```
package org.apache.spark.variant;
```

Since this is defined as `Variant represents a type that contain one of`, `org.apache.spark.type.*`, maybe?

```
package org.apache.spark.type.variant;
```

**Comment by cloud-fan** [comment]
SGTM, @chenhao-db can you open a followup PR to address it?

**Comment by dongjoon-hyun** [comment]
Thank you, @cloud-fan .

**Comment by dongjoon-hyun** [comment]
I made a PR, @cloud-fan and @chenhao-db .
- https://github.com/apache/spark/pull/45685

**Pull request closed without merging**